version: '3'

services:
  traefik:
    image: traefik:2.1.3
    restart: unless-stopped
    command:
      - --entrypoints.https.address=:443
      - --providers.docker=true
      - --log.level=INFO
      - --api.dashboard=true
      - --serverstransport.insecureskipverify=true
      - --accesslog.format=true
      - --certificatesresolvers.le.acme.email=${LETSENCRYPT_EMAIL}
      - --certificatesresolvers.le.acme.storage=/acme.json
      - --certificatesresolvers.le.acme.tlschallenge=true
    ports:
      - "443:443"
    volumes:
      # FIXME: make sure you run the following 2 commands *before*
      # starting this service (otherwise compose insists on creating
      # acme.json as a directory):
      #   touch ./docker-volumes/acme.json
      #   chmod 600 ./docker-volumes/acme.json
      - ./docker-volumes/acme.json:/acme.json
      - /var/run/docker.sock:/var/run/docker.sock:ro
    networks:
      - proxy-tier
    labels:
      traefik.http.routers.api.rule: Host(`${TRAEFIK_VIRTUAL_HOST}`)
      traefik.http.routers.api.service: api@internal
      traefik.http.routers.api.entrypoints: https
      traefik.http.routers.api.tls: true
      traefik.http.routers.api.tls.certresolver: le
      traefik.http.routers.api.middlewares: auth
      traefik.http.middlewares.auth.basicauth.users: ${TRAEFIK_AUTH_USER}

    logging:
      driver: fluentd
      options:
        tag: nextcloud.proxy
        fluentd-async-connect: 'true'
        fluentd-retry-wait: '1s'
        fluentd-max-retries: '30'

networks:
  proxy-tier:
